gprofiler/profilers/node.py

```diff
@@ -0,0 +1,244 @@
+    node_version = get_exe_version(process, Event(), 3)
+    # i. e. v16.3.2 -> 16
+    return node_version[1:].split(".")[0]
+
+
+@lru_cache(maxsize=1)
+def _get_dso_git_rev() -> str:
+    libc_dso_version_file = resource_path(os.path.join("node", "module", "glibc", "version"))
+    musl_dso_version_file = resource_path(os.path.join("node", "module", "musl", "version"))
+    with open(libc_dso_version_file) as f:
+        libc_dso_ver = f.readline()
+    with open(musl_dso_version_file) as f:
+        musl_dso_ver = f.readline()
+    # with no build errors this should always be the same
+    assert libc_dso_ver == musl_dso_ver
+    return libc_dso_ver
+
+
+@lru_cache()
+def _get_dest_inside_container(musl: bool, node_version: str) -> str:
+    libc = "musl" if musl else "glibc"
+    return os.path.join(TEMPORARY_STORAGE_PATH, "node_module", _get_dso_git_rev(), libc, node_version)
+
+
+def _start_debugger(pid: int) -> None:
+    # for windows: in shell node -e "process._debugProcess(PID)"
+    os.kill(pid, signal.SIGUSR1)
+
+
+@retry(NodeDebuggerUrlNotFound, 5, 1)
+def _get_debugger_url() -> str:
+    # when killing process with SIGUSR1 it will open new debugger session on port 9229,
+    # so it will always the same. When another debugger is opened in same NS it will not open new one.
+    # REF: Inspector agent initialization uses host_port
+    # https://github.com/nodejs/node/blob/5fad0b93667ffc6e4def52996b9529ac99b26319/src/inspector_agent.cc#L668
+    # host_port defaults to 9229
+    # ref: https://github.com/nodejs/node/blob/2849283c4cebbfbf523cc24303941dc36df9332f/src/node_options.h#L90
+    # in our case it won't be changed
+    port = 9229
+    debugger_url_response = requests.get(f"http://127.0.0.1:{port}/json/list")
+    if debugger_url_response.status_code != 200 or "application/json" not in debugger_url_response.headers.get(
+        "Content-Type", ""
+    ):
+        raise NodeDebuggerUrlNotFound(
+            {"status_code": debugger_url_response.status_code, "text": debugger_url_response.text}
+        )
+
+    response_json = debugger_url_response.json()
+    if (
+        not isinstance(response_json, list)
+        or len(response_json) == 0
+        or not isinstance(response_json[0], dict)
+        or "webSocketDebuggerUrl" not in response_json[0]
+    ):
+        raise NodeDebuggerUrlNotFound(response_json)
+
+    return cast(str, response_json[0]["webSocketDebuggerUrl"])
+
+
+def _send_socket_request(sock: WebSocket, cdp_request: Dict) -> None:
+    sock.send(json.dumps(cdp_request))
+    message = sock.recv()
+    try:
+        message = json.loads(message)
+    except json.JSONDecodeError:
+        raise NodeDebuggerUnexpectedResponse(message)
+
+    if (
+        "result" not in message.keys()
+        or "result" not in message["result"].keys()
+        or "type" not in message["result"]["result"].keys()
+        or message["result"]["result"]["type"] != "boolean"
```

Comment:
No further checks on the results need to be done here, btw? (I figured, if you check that the type is `boolean`, then we should probably check that the value is `true`, no?)

Reply:
Checking value will only show us if module was started/stopped before, which isn't necessary in my opinion. Seeing it returned boolean lets us know, that module is now in wanted state.

Reply:
I see. Since the module cannot fail (and return value is only "did I turn it on/off now"), you're right